### PR TITLE
[n-beats] keras model implementation

### PR DESCRIPTION
**Pull request by eljdos**
Please do find a keras implementation from your work.

User could create a results/ repo, a results/test/ repo and a data/train and data/test repo.
Within the data repo,, user could put the M4 data (https://www.mcompetitions.unic.ac.cy/the-dataset/)

### CODE REVIEW

File: n_beats/data/val/Daily-test.csv [hunk 1]
```diff
@@ -0,0 +1,100 @@
+D1,2039.2,2035,2051.8,2061.8,2063.5,2069.5,2054,2057,2062.8,2066.4,2067.4,2071.4,2083.8,2080.6
```
Comment (philipperemy) [inline n_beats/data/val/Daily-test.csv:1]: I think you forgot the header here.

File: n_beats/model_keras.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+import matplotlib.pyplot as plt
+import numpy as np
+import tensorflow as tf
+from keras import backend as K
+from keras.layers import Input, Dense, Lambda, Subtract, Add
+from keras.models import Model
+from keras.optimizers import Adam
+
+
+class N_Beats:
```
Comment (philipperemy) [inline n_beats/model_keras.py]: Python: Should be NBeats.

File: n_beats/model_keras.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+                d1 = Dense(self.units, activation='relu')(x_)
+                d2 = Dense(self.units, activation='relu')(d1)
+                d3 = Dense(self.units, activation='relu')(d2)
+                d4 = Dense(self.units, activation='relu')(d3)
+                if self.block_types[j] == 'generic':
+                    theta_b = Dense(self.nb_poly, activation='linear')(d4)
+                    theta_f = Dense(self.nb_poly, activation='linear')(d4)
+                    backcast = Dense(self.backcast_length, activation='linear')(theta_b)
+                    forecast = Dense(self.forecast_length, activation='linear')(theta_f)
+                    x_ = Subtract()([x_, backcast])
```
Comment (philipperemy) [inline n_beats/model_keras.py]: This part is common to the 3 blocks. You can factorize it.
```
                    x_ = Subtract()([x_, backcast])
                    if i == 0 and j == 0:
                        y_ = forecast
                    else :
                        y_ = Add()([y_, forecast])
````

File: n_beats/model_keras.py [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+                    if i == 0 and j == 0:
+                        y_ = forecast
+                    else :
+                        y_ = Add()([y_, forecast])
+
+        model = Model(X_, y_)
+        model.summary()
+
+        optimizer = Adam(lr=0.00001)
```
Comment (philipperemy) [inline n_beats/model_keras.py]: Technically you should put this in another function.

```
def compile(loss, learning_rate=1e-5):
     optimizer = Adam(lr=learning_rate)
     self.nbeats.compile(loss=loss, optimizer=optimizer)
```

File: n_beats/model_keras.py [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+        self.nbeats = model
+        self.nbeats.compile(loss='mae', optimizer=optimizer)
+
+    def train_model(self):
+
+        print('loading dev data')
+        x_test, y_test = get_xtest_circular(self.backcast_length, self.forecast_length)
+        print(x_test.shape)
```
Comment (philipperemy) [inline n_beats/model_keras.py]: For every print, check if it's necessary.
You can add `print('x_test.shape=', x_test.shape)`

File: n_beats/model_keras.py [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+                print(step)
+                if not os.path.exists('results'):
+                    os.makedirs('results')
+                self.nbeats.save('results/n_beats_model_20_' + str(step) + '.h5')
+                predictions = self.nbeats.predict(x_test)
+                residual = y_test - predictions
+                print(get_accuracy(y_test, predictions))
+                if get_accuracy(y_test, predictions)[0] < self.best_perf:
+                    self.best_perf = get_accuracy(y_test, predictions)[0]
+                    self.nbeats.save('results/n_beats_model_20_ongoing.h5')
```
Comment (philipperemy) [inline n_beats/model_keras.py]: why 20?

File: n_beats/model_keras.py [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+                print(residual.shape)
+                plt.figure()
+                plt.grid(True)
+                print(x.shape)
+                print(y.shape)
+                print(predictions.shape)
+                print(residual.shape)
+                x_y = np.concatenate([x[0, :], y[0, :]], axis=-1).flatten()
+                plt.plot(list(range(self.backcast_length)), x[0, :].flatten(), color='b')
+                plt.plot(list(range(len(x_y) - self.forecast_length, len(x_y))), y[0, :].flatten(), color='g')
```
Comment (philipperemy) [inline n_beats/model_keras.py]: Could store `range(len(x_y) - self.forecast_length, len(x_y))` in a variable

File: n_beats/model_keras.py [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+                plt.legend(['backcast', 'forecast', 'predictions of forecast'])
+                if not os.path.exists('results/test'):
+                    os.makedirs('results/test')
+                plt.savefig('results/test/' + str(step) + '.png')
+                plt.close()
+
+        self.nbeats.save('results/n_beats_model_20_small.h5')
+
+        predictions = self.nbeats.predict(x)
+        x_y = np.concatenate([x[100, :], y[100, :]], axis=-1).flatten()
```
Comment (philipperemy) [inline n_beats/model_keras.py]: Not a big deal but I feel you could factorize this block in a function as it is used many times in the code. It would improve readability. Especially some have plt.grid(True) and some not.

File: n_beats/model_keras.py [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+                residual = y_test - predictions
+                print(get_accuracy(y_test, predictions))
+                if get_accuracy(y_test, predictions)[0] < self.best_perf:
+                    self.best_perf = get_accuracy(y_test, predictions)[0]
+                    self.nbeats.save('results/n_beats_model_20_ongoing.h5')
+                print(predictions.shape)
+                print(residual.shape)
+                plt.figure()
+                plt.grid(True)
+                print(x.shape)
```
Comment (philipperemy) [inline n_beats/model_keras.py]: Print not necessary. Or you can maybe add them on the same line :)

File: n_beats/model_keras.py [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+        
+    ts_min = np.min(x_test, axis = 1).reshape((x_test.shape[0], 1))
+    ts_max = np.max(x_test, axis = 1).reshape((x_test.shape[0], 1))
+    x_test = (x_test - ts_min) / (ts_max - ts_min + 1e-8)
+    y_test = (y_test - ts_min) / (ts_max - ts_min + 1e-8)
+
+    return x_test, y_test
+
+
+def get_accuracy(y_true, y_hat):
```
Comment (philipperemy) [inline n_beats/model_keras.py]: Technically an accuracy is for a classification problem. Here we deal with regression.

File: n_beats/model_keras.py [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+
+        print(get_accuracy(y_test, validation_predictions))
+
+
+def linear_space(length, fwd_looking=True):
+    if fwd_looking:
+        t = np.linspace(0, length - 1, length)
+    else:
+        t = np.linspace(-length, -1, length)
+    t = t / length
```
Comment (philipperemy) [inline n_beats/model_keras.py]: Maybe mistake here: (to be checked!)
If forecast_length=5, linear space = [0.  0.2 0.4 0.6 0.8]
and backcast_length=10, linear space = [-1.  -0.9 -0.8 -0.7 -0.6 -0.5 -0.4 -0.3 -0.2 -0.1]
I'm not sure but I think the linspace step should be the same for the backcast and the forecast.
For example [-1.  -0.9 -0.8 -0.7 -0.6 -0.5 -0.4 -0.3 -0.2 -0.1] is OK
Then forecast [0, 0.1, 0.2, 0.3, 0.4].

File: n_beats/model_keras.py [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+    s2 = np.stack([np.sin(2 * np.pi * i * t) for i in range(p2)], axis=0)
+    S = np.concatenate([s1, s2], axis=0)
+    S = K.cast(S, np.float32)
+    return K.dot(thetas, S)
+
+
+def trend_model(thetas, length, is_forecast):
+    p = thetas.shape[-1]
+    t = linear_space(length, fwd_looking=is_forecast)
+    T = K.transpose(tf.stack([t ** i for i in range(p)], axis=0))
```
Comment (philipperemy) [inline n_beats/model_keras.py]: `K.stack` instead of `tf.stack`?

File: n_beats/model_keras.py [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+        t = np.linspace(-length, -1, length)
+    t = t / length
+    return t
+
+
+def seasonality_model(thetas, length, is_forecast):
+    p = thetas.shape[-1]
+    p1, p2 = (p // 2, p // 2) if p % 2 == 0 else (p // 2, p // 2 + 1)
+    t = linear_space(length, fwd_looking=is_forecast)
+    s1 = np.stack([np.cos(2 * np.pi * i * t) for i in range(p1)], axis=0)  # H/2-1
```
Comment (philipperemy) [inline n_beats/model_keras.py]: you can use tf.cos instead of np.cos for all.

Something like this would work (to be checked!)

```
p = thetas.get_shape().as_list()[-1]
p1, p2 = (p // 2, p // 2) if p % 2 == 0 else (p // 2, p // 2 + 1)
s1 = tf.stack([tf.cos(2 * np.pi * i * t) for i in range(p1)], axis=0)  # H/2-1
s2 = tf.stack([tf.sin(2 * np.pi * i * t) for i in range(p2)], axis=0)
 S = tf.concat([s1, s2], axis=0)
return tf.matmul(thetas, S)
```

File: n_beats/model_keras.py [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+
+
+def train():
+    nbeats = N_Beats()
+    print(nbeats)
+    nbeats.train_model()
+
+
+if __name__ == '__main__':
+    train()
```
Comment (philipperemy) [inline n_beats/model_keras.py]: no white space at the end of file.

File: n_beats/model_keras.py (hunk 11 above)
Comment (philipperemy) [reply n_beats/model_keras.py]: Could be as simple as (again to be checked, just an idea)!

```
l = tf.linspace(-float(backcast_length), float(forecast_length),
                    backcast_length + forecast_length) / forecast_length
    l_backcast = l[:backcast_length]
    l_forecast = l[backcast_length:]
```

File: n_beats/model_keras.py [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+        y_train = np.vstack((y_train, time_series_cleaned_forlearning_y))
+
+    ts_min = np.min(x_train, axis = 1).reshape((x_train.shape[0], 1))
+    ts_max = np.max(x_train, axis = 1).reshape((x_train.shape[0], 1))
+    x_train = (x_train - ts_min) / (ts_max - ts_min + 1e-8)
+    y_train = (y_train - ts_min) / (ts_max - ts_min + 1e-8)
+    return x_train, y_train
+
+
+def get_xtest_circular(backcast_length, forecast_length):
```
Comment (philipperemy) [inline n_beats/model_keras.py]: More nice to have: Could `get_xtrain_circular` and `get_xtest_circular` be just one function? But really not necessary here.

File: n_beats/model_keras.py [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+            (time_series_cleaned.shape[0] + 1 - (backcast_length + forecast_length), forecast_length))
+        for j in range(backcast_length, time_series_cleaned.shape[0] + 1 - forecast_length):
+            time_series_cleaned_forlearning_x[j - backcast_length, :] = time_series_cleaned[j - backcast_length:j]
+            time_series_cleaned_forlearning_y[j - backcast_length, :] = time_series_cleaned[j: j + forecast_length]
+        x_test = np.vstack((x_test, time_series_cleaned_forlearning_x))
+        y_test = np.vstack((y_test, time_series_cleaned_forlearning_y))
+        
+    ts_min = np.min(x_test, axis = 1).reshape((x_test.shape[0], 1))
+    ts_max = np.max(x_test, axis = 1).reshape((x_test.shape[0], 1))
+    x_test = (x_test - ts_min) / (ts_max - ts_min + 1e-8)
```
Comment (philipperemy) [inline n_beats/model_keras.py]: Technically a normalization of inputs using `ts_max` or `ts_min` is forward looking meaning that the model can see information from the future. It's not a big deal to show how to use the model though :)

File: n_beats/model_keras.py [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+    error = np.mean(np.square(true_y - preds))
+    smape = np.mean(2 * np.abs(true_y - preds) / (np.abs(true_y) + np.abs(preds)))
+    correl = np.corrcoef(true_y, preds)
+
+    return smape, error, correl
+
+
+def train():
+    nbeats = N_Beats()
+    print(nbeats)
```
Comment (philipperemy) [inline n_beats/model_keras.py]: If you want to print nbeats, maybe define the function `__str__` for NBeats()

File: n_beats/data/val/Daily-test.csv (hunk 1 above)
Comment (eljdos) [reply n_beats/data/val/Daily-test.csv:1]: I think the header is here.

File: n_beats/model_keras.py (hunk 2 above)
Comment (eljdos) [reply n_beats/model_keras.py]: Corrected

File: n_beats/model_keras.py (hunk 3 above)
Comment (eljdos) [reply n_beats/model_keras.py]: Corrected

File: n_beats/model_keras.py (hunk 4 above)
Comment (eljdos) [reply n_beats/model_keras.py]: Updated

File: n_beats/model_keras.py (hunk 5 above)
Comment (eljdos) [reply n_beats/model_keras.py]: Corrected

File: n_beats/model_keras.py (hunk 6 above)
Comment (eljdos) [reply n_beats/model_keras.py]: Corrected

File: n_beats/model_keras.py (hunk 8 above)
Comment (eljdos) [reply n_beats/model_keras.py]: Updated

File: n_beats/model_keras.py (hunk 9 above)
Comment (eljdos) [reply n_beats/model_keras.py]: Corrected

File: n_beats/model_keras.py (hunk 10 above)
Comment (eljdos) [reply n_beats/model_keras.py]: Updated

File: n_beats/model_keras.py (hunk 11 above)
Comment (eljdos) [reply n_beats/model_keras.py]: That does sound better :)

File: n_beats/model_keras.py (hunk 12 above)
Comment (eljdos) [reply n_beats/model_keras.py]: Corrected

File: n_beats/model_keras.py (hunk 13 above)
Comment (eljdos) [reply n_beats/model_keras.py]: Corrected

File: n_beats/model_keras.py (hunk 14 above)
Comment (eljdos) [reply n_beats/model_keras.py]: Sorry !

File: n_beats/model_keras.py (hunk 15 above)
Comment (eljdos) [reply n_beats/model_keras.py]: I tried to factorize a bit

File: n_beats/model_keras.py (hunk 16 above)
Comment (eljdos) [reply n_beats/model_keras.py]: I took it off :)

File: n_beats/model_keras.py (hunk 17 above)
Comment (eljdos) [reply n_beats/model_keras.py]: I'm not sure it's needed here so I just took it off.

**Comment by philipperemy** [comment]
Thank you!!
